2015-10-31-dekorator.md

```diff
@@ -0,0 +1,181 @@
+---
+title: Dekorator
+description: >
+    Wzorzec dekorator stosowany jest tam, gdzie dziedziczenie nie jest optymalnym sposobem rozszerzania funkcjonalności
+    klasy. Polega na dołączeniu nowych atrybutów poprzez “opakowanie” obiektu bazowego obiektem zwanym dekoratorem.
+enabled: true
+type: Strukturalny 
+usage_info: null
+posts: []
+
+---
+{% block description %}
+Wyobraźmy sobie sytuację, gdzie kombinacja wielu dodatków do bazowej klasy spowodowałaby rozlew klas dziedziczących,
+które programista musiałby zaimplementować w aplikacji, na przykład:
+PizzaWithPepperoniAndTabasco extends Pizza, PizzaWithPepperoniAndMushrooms extends Pizza i tak dalej. Można sobie łatwo
+wyobrazić jak wiele typowa pizzeria może stworzyć kombinacji tego smacznego placka. Z pomocą przychodzi wzorzec
+dekorator, który umożliwia nam rozszerzanie funkcjonalności obiektu w czasie pracy aplikacji. 
+{% endblock %}
+
+{% block dummy_implementation_php %}
+```language-php
+interface SomeFancyInterface
+{
+    public function getInfo();
+}
+
+class BasicObject implements SomeFancyInterface
+{
+    public function getInfo()
+    {
+        return "I'm basic object";
+    }
+}
+
+class Decorator implements SomeFancyInterface
+{
+    private $object;
+
+    public function __construct(SomeFancyInterface $object)
+    {
+        $this->object = $object;
+    }
+
+    public function getInfo()
+    {
+        return $this->object->getInfo() . " decorated!";
+    }
+}
+
+$object = new BasicObject();
+$object = new Decorator($object);
```

Comment:
To zapisz jako
```php
$object = new Decorator(new BasicObject());
```
nadpisywanie zmiennej, nie jest dobrym pomysłem nawet jeżeli idzie tylko o zaprezentowanie idei